### PR TITLE
[Projects] Add RemoteRunner for supporting pipeline scheduling

### PR DESCRIPTION
**Pull request by yonishelach**
preview:
- Created `_RemoteRunner` engine
- Created `load_and_run()` and added it to `mlrun/projects/__init__.py`
- Added `schedule` parameter to `WorkflowSpec` and to `project.run()`
- Moved `get_run_status()` from `mlrun/projects/project.py` to `_PipelineRunner` class in `mlrun/projects/pipelines.py`
- Removed `get_run_status()` from `__main__.py`
- Created system tests to remote engine

Notes:
- Need to merge [project-demo-10](https://github.com/mlrun/project-demo/pull/10) in mlrun/project-demo for supporting the new system tests.
- Setting `watch = False` with `engine = remote:local` does not work due to `_LocalRunner.run()` implementation.

### CODE REVIEW

File: mlrun/projects/pipelines.py [hunk 1]
```diff
@@ -557,6 +564,72 @@ def run(
...
+        artifact_path=None,
+        namespace=None,
+        schedule=None,
+    ) -> _PipelineRunStatus:
+        workflow_id = uuid.uuid4().hex
+        logger.info(workflow_spec.to_dict())
+        try:
+            logger.info('Creating the function that invokes the workflow remotely')
+            fn = mlrun.new_function(
+                name=f'remote_runner_{workflow_id}',
```
Comment (yaronha) [inline mlrun/projects/pipelines.py:594]: no need for function per workflow (its the same one), use `workflow_runner`

File: mlrun/projects/pipelines.py [hunk 2]
```diff
@@ -557,6 +564,72 @@ def run(
...
+            )
+            logger.info('Running the function that invokes the workflow remotely')
+            # Preparing parameters for load_and_run function:
+            params = workflow_spec.args.copy() if workflow_spec.args else {}
+            params['workflow_name'] = name.split('-')[-1] if f'{project.name}-' in name else name
+            params['local'] = workflow_spec.run_local
+
+            fn.run(
+                params=params,
+                handler=mlrun.projects.pipelines.load_and_run,
```
Comment (yaronha) [inline mlrun/projects/pipelines.py:608]: should be a string: "mlrun.projects.pipelines.load_and_run"

File: mlrun/projects/pipelines.py [hunk 3]
```diff
@@ -38,11 +38,15 @@ def get_workflow_engine(engine_kind, local=False):
             logger.warning(
                 "running kubeflow pipeline locally, note some ops may not run locally!"
             )
+        elif engine_kind == 'remote':
```
Comment (yaronha) [inline mlrun/projects/pipelines.py:41]: No, this section is only for (force) `local` runs

File: mlrun/projects/project.py [hunk 4]
```diff
@@ -1907,15 +1921,26 @@ def run(
         if workflow_path or (workflow_handler and callable(workflow_handler)):
             workflow_spec = WorkflowSpec(path=workflow_path, args=arguments)
         else:
-            workflow_spec = WorkflowSpec.from_dict(self.spec._workflows[name])
+            workflow_spec = self.spec._workflows[name].copy()
```
Comment (guy1992l) [inline mlrun/projects/project.py:1924]: Keep in mind that `copy` do not copy lists and dicts. If the spec has those kind of field its better to use deepcopy.

File: tests/system/projects/test_project.py [hunk 5]
```diff
@@ -288,6 +288,58 @@ def test_local_pipeline(self):
...
         self._test_new_pipeline("kfppipe", engine="kfp")
 
+    def _test_remote_pipeline_from_github(
+        self, name, workflow_name, engine=None, local=None, watch=False
+    ):
+        project_dir = f"{projects_dir}/{name}"
+        shutil.rmtree(project_dir, ignore_errors=True)
+        # TODO: Need to replace with git://github.com/mlrun/project-demo.git
+        project = mlrun.load_project(
+            project_dir, "git://github.com/yonishelach/project-demo.git", name=name
```
Comment (guy1992l) [inline tests/system/projects/test_project.py:298]: same here

**Review by guy1992l — CHANGES_REQUESTED** [review]
Great job Yoni! Seems like you learned a lot!

The major edits I wish for you to do is some cosmetics:
* Add captions `#` that explain what you do and why.
* Add type hints for attributes and return values.
* Add doc strings for user oriented functions.

The other stuff are some notes for you to keep in mind, like the copy or deepcopy, abstract functions implementations and optional / default attributes in functions.

File: mlrun/projects/pipelines.py [hunk 6]
```diff
@@ -598,3 +797,57 @@ def github_webhook(request):
...
+    namespace=None,
+    sync: bool = False,
+    watch: bool = False,
+    dirty: bool = False,
+    ttl=None,
+    engine=None,
+    local: bool = None,
+    timeout: int = None,
+):
+    # Loading the project:
```
[the remark this answers is not in the document]
Comment (yonishelach) [reply mlrun/projects/pipelines.py:823]: I will add type hints, but actually this function is for internal use, so maybe it is best to change the name of this function to `_load_and_run`.

File: mlrun/projects/pipelines.py [hunk 7]
```diff
@@ -504,6 +536,49 @@ def get_state(run_id, project=None):
...
+    @staticmethod
+    def get_run_status(
+        project,
+        run,
+        timeout=None,
+        expected_statuses=None,
+        notifiers: RunNotifications = None,
+    ):
+        if timeout is None:
+            timeout = 60 * 60
```
[the remark this answers is not in the document]
Comment (yonishelach) [reply mlrun/projects/pipelines.py:548]: @Tankilevitch did it this way, can you elaborate?

File: mlrun/projects/pipelines.py [hunk 8]
```diff
@@ -343,12 +358,15 @@ def enrich_function_object(
...
-    def __init__(self, run_id, engine, project, workflow=None, state=""):
+    def __init__(
+        self, run_id, engine, project, workflow=None, state="", run_object=None
+    ):
         self.run_id = run_id
         self.project = project
         self.workflow = workflow
         self._engine = engine
         self._state = state
+        self.run_object = run_object  # for _RemoteRunner
```
Comment (yaronha) [inline mlrun/projects/pipelines.py:369]: should be private (start with `_`)

File: mlrun/projects/pipelines.py [hunk 9]
```diff
@@ -457,6 +490,7 @@ def run(
         secrets=None,
         artifact_path=None,
         namespace=None,
+        schedule=None,
```
Comment (yaronha) [inline mlrun/projects/pipelines.py:493]: why is that here and not in the `workflow_spec` ?

File: mlrun/projects/pipelines.py [hunk 10]
```diff
@@ -520,6 +599,7 @@ def run(
...
```
Comment (yaronha) [inline mlrun/projects/pipelines.py:602]: same

File: mlrun/projects/pipelines.py [hunk 11]
```diff
@@ -557,6 +636,141 @@ def run(
...
+    def run(
+        cls,
+        project,
+        workflow_spec: WorkflowSpec,
+        name=None,
+        workflow_handler=None,
+        secrets=None,
+        artifact_path=None,
+        namespace=None,
+        schedule=None,
```
Comment (yaronha) [inline mlrun/projects/pipelines.py:671]: same, should be in spec

File: mlrun/projects/pipelines.py [hunk 12]
```diff
@@ -557,6 +636,141 @@ def run(
...
+                    "artifact_path": artifact_path,
+                    "workflow_handler": workflow_handler or workflow_spec.handler,
+                    "namespace": namespace,
+                    "ttl": workflow_spec.ttl,
+                    "engine": workflow_spec.engine,
+                    "local": workflow_spec.run_local,
+                },
+                handler="mlrun.projects.load_and_run",
+                local=False,
+                schedule=schedule,
```
Comment (yaronha) [inline mlrun/projects/pipelines.py:709]: same

File: mlrun/__main__.py [hunk 13]
```diff
@@ -959,19 +959,19 @@ def project(
...
                 workflow_path,
                 arguments=args,
                 artifact_path=artifact_path,
                 namespace=namespace,
                 sync=sync,
+                watch=watch,
                 dirty=dirty,
                 workflow_handler=handler,
                 engine=engine,
                 local=local,
```
Comment (yaronha) [inline mlrun/__main__.py:973]: add schedule param

**Review by Tankilevitch — CHANGES_REQUESTED** [review]

File: mlrun/__main__.py [hunk 14]
```diff
@@ -878,6 +878,13 @@ def logs(uid, project, offset, db, watch):
     is_flag=True,
     help="ensure the project exists, if not, create project",
 )
+@click.option(
+    "--schedule",
+    type=str,
+    default=None,
+    help="a standard crontab expression string, for help see: "
+    "https://apscheduler.readthedocs.io/en/3.x/modules/triggers/cron.html#module-apscheduler.triggers.cron",
+)
```
[the remark this answers is not in the document]
Comment (yonishelach) [reply mlrun/__main__.py]: done

File: mlrun/projects/pipelines.py [hunk 15]
```diff
@@ -598,3 +809,53 @@ def github_webhook(request):
...
+    workflow_path: typing.Optional[str] = None,
+    workflow_arguments: typing.Optional[dict] = None,
+    artifact_path: typing.Optional[str] = None,
+    workflow_handler: typing.Optional[str] = None,
+    namespace: typing.Optional[str] = None,
+    sync: bool = False,
+    dirty: bool = False,
+    ttl: typing.Optional[int] = None,
+    engine: typing.Optional[str] = None,
+    local: typing.Optional[bool] = None,
```
[the remark this answers is not in the document]
Comment (yonishelach) [reply mlrun/projects/pipelines.py]: done

File: mlrun/projects/project.py [hunk 16]
```diff
@@ -608,33 +608,45 @@ def remove_function(self, name):
...
+                )
+            if isinstance(workflow, dict):
+                workflow = WorkflowSpec.from_dict(workflow)
+            name = workflow.name
             # todo: support steps dsl as code alternative
             if not name:
                 raise ValueError('workflow "name" must be specified')
-            if "path" not in workflow and "code" not in workflow:
+            if not workflow.path and not workflow.code:
                 raise ValueError('workflow source "path" or "code" must be specified')
```
[the remark this answers is not in the document]
Comment (yonishelach) [reply mlrun/projects/project.py:639]: @Tankilevitch need at least one of `path`, `code` here so it is still need to be **or**

File: mlrun/projects/project.py [hunk 17]
```diff
@@ -1907,15 +1931,28 @@ def run(
...
+        if schedule is not None:
+            workflow_spec.schedule = schedule
+
+        inner_engine = None
+        if engine and engine.startswith("remote"):
+            if ":" in engine:
+                engine, inner_engine = engine.split(":")
+        elif workflow_spec.schedule:
+            inner_engine = engine
+            engine = "remote"
```
[the remark this answers is not in the document]
Comment (yonishelach) [reply mlrun/projects/project.py:1951]: There are the following cases:
* if `schedule` is given without `engine` then the main engine will be `remote` and the inner engine will be inferred or will be set to the given one.
* if `engine` is `remote` then the inner engine will be inferred.
* if `engine` is `remote:X` then the main engine is `remote` and the inner is `X`.

File: mlrun/projects/project.py [hunk 18]
```diff
@@ -1907,15 +1931,28 @@ def run(
...
         workflow_engine = get_workflow_engine(engine or workflow_spec.engine, local)
-        workflow_spec.engine = workflow_engine.engine
+        if not inner_engine and engine == "remote":
+            inner_engine = get_workflow_engine(workflow_spec.engine, local).engine
+        workflow_spec.engine = inner_engine or workflow_engine.engine
```
[the remark this answers is not in the document]
Comment (yonishelach) [reply mlrun/projects/project.py:1956]: The default is kfp for the remote case also, added comment anyway

File: mlrun/projects/project.py [hunk 19]
```diff
@@ -1926,9 +1963,14 @@ def run(
             artifact_path=artifact_path,
             namespace=namespace,
         )
+        run_msg = f"started run workflow {name} "
+        if run and hasattr(run, "run_id"):
+            run_msg += f"with run id = '{run.run_id}' "
```
[the remark this answers is not in the document]
Comment (yonishelach) [reply mlrun/projects/project.py]: you right, edited + added comment

File: mlrun/projects/project.py [hunk 20]
```diff
@@ -608,33 +608,45 @@ def remove_function(self, name):
...
 
     @property
-    def workflows(self) -> list:
-        """list of workflows specs used in this project"""
-        return [workflow for workflow in self._workflows.values()]
+    def workflows(self) -> typing.List[dict]:
+        """
+        :returns: list of workflows specs dicts used in this project
+        """
+        return [workflow.to_dict() for workflow in self._workflows.values()]
```
[the remark this answers is not in the document]
Comment (yonishelach) [reply mlrun/projects/project.py:615]: meanwhile, `WorkflowSpec` is for internal use

File: mlrun/__main__.py [hunk 21]
```diff
@@ -902,6 +909,7 @@ def project(
     env_file,
     timeout,
     ensure_project,
+    schedule,
```
Comment (Hedingber) [inline mlrun/__main__.py:912]: If the flag name is `--schedule-pipeline` I think the variable name should be `schedule_pipeline` meaning this is not really working, you had a lot of comments on this PR, meaning a lot of changes, please make sure to test everything after all changes - ideally obviously to just write tests so it's automatic

**Review by Hedingber — CHANGES_REQUESTED** [review]
The code here basically looks good to me, but design-wise, I don't really understand why in order to schedule pipelines we need to bring up a job.. (and in addition I think this whole `remote` engine thing won't be clear to the user)
Why aren't we just using our scheduler ? it already has a `kind` parameter, which even already has a `pipeline` in the enum which was added back then as "hachana-lemazgan", bringing a job for it every time sounds redundant to me, but maybe I'm missing the value it gives

**Comment by yaronha** [comment]
> I don't really understand why in order to schedule pipelines we need to bring up a job.. 

@Hedingber the Jon is not just running a function, it loads the archive (from git/tar/..) and can execute tasks locally, we don't want to do this in the MLRun service, but rather on a separate POD. We also want this to appear as a managed scheduled task in the UI w option to delete/etc. (at some point the UI will also see the label a give it a sched workflow icon)

**Comment by yaronha** [comment]
> > I don't really understand why in order to schedule pipelines we need to bring up a job..
> 
> @Hedingber the Jon is not just running a function, it loads the archive (from git/tar/..) and can execute tasks locally, we don't want to do this in the MLRun service, but rather on a separate POD. We also want this to appear as a managed scheduled task in the UI w option to delete/etc. (at some point the UI will also see the label a give it a sched workflow icon)

BTW it can be a good idea to set the sched job type to `pipeline` (use the enum) for the UI/API , may require propagating this via the run/submit command..

File: mlrun/__main__.py (hunk 21 above)
Comment (yaronha) [reply mlrun/__main__.py:912]: i think it should be `--schedule` since its obvious we talk about a pipeline

Comment (Tankilevitch) [reply mlrun/__main__.py:912]: lets keep it`--schedule` this function is for pipelines

Comment (yonishelach) [reply mlrun/__main__.py:912]: back to `--schedule`

File: mlrun/projects/pipelines.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (Tankilevitch) [reply mlrun/projects/pipelines.py:551]: we had a bug in which when user passed timeout=None the default which was 60*60 have been overridden.

**Review by Tankilevitch — CHANGES_REQUESTED** [review]

File: tests/system/projects/test_project.py [hunk 22]
```diff
@@ -288,6 +288,62 @@ def test_local_pipeline(self):
...
+        )
+        try:
+            self._test_remote_pipeline_from_github(
+                name="rmtpipe-local-github",
+                workflow_name="newflow",
+                engine="remote",
+                local=True,
+            )
+        except Exception as e:
+            assert isinstance(e, mlrun.errors.MLRunInvalidArgumentError), str(e)
```
Comment (Tankilevitch) [inline tests/system/projects/test_project.py:345]:
```suggestion
        with pytests.raises(mlrun.errors.MLRunInvalidArgumentError):
            self._test_remote_pipeline_from_github(
                name="rmtpipe-local-github",
                workflow_name="newflow",
                engine="remote",
                local=True,
            )
```

File: tests/system/projects/test_project.py [hunk 23]
```diff
@@ -288,6 +288,62 @@ def test_local_pipeline(self):
...
+            local=local,
+            engine=engine,
+        )
+
+        assert run.state == mlrun.run.RunStatuses.succeeded, "pipeline failed"
+        assert run.run_id, "workflow's run id failed to fetch"
+        assert (
+            workflow_dict.items() <= run.workflow.to_dict().items()
+        ), "workflow executed with wrong parameters"
+        self._delete_test_project()
```
Comment (Tankilevitch) [inline tests/system/projects/test_project.py:317]: is this required in your other tests as well?

**Review by Tankilevitch — CHANGES_REQUESTED** [review]
Looks good, really last mile changes
Also I think you are missing tests like running that from CLI like in `test_run_cli`.

File: tests/system/projects/test_project.py (hunk 23 above)
Comment (yonishelach) [reply tests/system/projects/test_project.py:345]: removed.. not really required

File: tests/system/projects/test_project.py (hunk 22 above)
Comment (yonishelach) [reply tests/system/projects/test_project.py]: done

**Review by Tankilevitch — APPROVED** [review]
